core/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -592,6 +592,11 @@ public synchronized void updateMetaData(final IndexMetaData metadata) {
                 rescheduleFsyncTask(durability);
             }
         }
+
+        // update primary terms
+        for (final IndexShard shard : this.shards.values()) {
+            shard.updatePrimaryTerm(metadata.primaryTerm(shard.shardId().id()));
```

Comment:
We can push this down to `IndexShard.onSettingsChanged()`. No need to touch `IndexService`.